pkg/webhook/validate.go

```diff
@@ -189,98 +185,105 @@ func dynamicValidateSriovNetworkNodePolicy(cr *sriovnetworkv1.SriovNetworkNodePo
 	}
-	if !interfaceSelected {
-		return false, fmt.Errorf("no supported NIC is selected by the nicSelector in CR %s", cr.GetName())
-	}
 
 	return true, nil
 }
 
-func validatePolicyForNodeState(policy *sriovnetworkv1.SriovNetworkNodePolicy, state *sriovnetworkv1.SriovNetworkNodeState, node *corev1.Node) (bool, error) {
+func validatePolicyForNodeStateAndPolicy(nsList *sriovnetworkv1.SriovNetworkNodeStateList, npList *sriovnetworkv1.SriovNetworkNodePolicyList, node *corev1.Node, cr *sriovnetworkv1.SriovNetworkNodePolicy) error {
+	for _, ns := range nsList.Items {
+		if ns.GetName() == node.GetName() {
+			if err := validatePolicyForNodeState(cr, &ns, node); err != nil {
+				return err
+			}
+		}
+	}
+	// validate current policy against policies in API (may not be converted to SriovNetworkNodeState yet)
+	for _, np := range npList.Items {
+		if np.GetName() != cr.GetName() && np.Selected(node) {
+			if err := validatePolicyForNodePolicy(cr, &np); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func validatePolicyForNodeState(policy *sriovnetworkv1.SriovNetworkNodePolicy, state *sriovnetworkv1.SriovNetworkNodeState, node *corev1.Node) error {
 	glog.V(2).Infof("validatePolicyForNodeState(): validate policy %s for node %s.", policy.GetName(), state.GetName())
 	for _, iface := range state.Status.Interfaces {
-		if validateNicModel(&policy.Spec.NicSelector, &iface, node) {
-			interfaceSelected = true
+		err := validateNicModel(&policy.Spec.NicSelector, &iface, node)
+		if err == nil {
 			if policy.GetName() != constants.DefaultPolicyName && policy.Spec.NumVfs == 0 {
-				return false, fmt.Errorf("numVfs(%d) in CR %s is not allowed", policy.Spec.NumVfs, policy.GetName())
+				return fmt.Errorf("numVfs(%d) in CR %s is not allowed", policy.Spec.NumVfs, policy.GetName())
 			}
 			if policy.Spec.NumVfs > iface.TotalVfs && iface.Vendor == IntelID {
-				return false, fmt.Errorf("numVfs(%d) in CR %s exceed the maximum allowed value(%d)", policy.Spec.NumVfs, policy.GetName(), iface.TotalVfs)
+				return fmt.Errorf("numVfs(%d) in CR %s exceed the maximum allowed value(%d)", policy.Spec.NumVfs, policy.GetName(), iface.TotalVfs)
 			}
 			if policy.Spec.NumVfs > MlxMaxVFs && iface.Vendor == MellanoxID {
-				return false, fmt.Errorf("numVfs(%d) in CR %s exceed the maximum allowed value(%d)", policy.Spec.NumVfs, policy.GetName(), MlxMaxVFs)
+				return fmt.Errorf("numVfs(%d) in CR %s exceed the maximum allowed value(%d)", policy.Spec.NumVfs, policy.GetName(), MlxMaxVFs)
 			}
 			// vdpa: only mellanox cards are supported
 			if policy.Spec.VdpaType == constants.VdpaTypeVirtio && iface.Vendor != MellanoxID {
-				return false, fmt.Errorf("vendor(%s) in CR %s not supported for virtio-vdpa", iface.Vendor, policy.GetName())
+				return fmt.Errorf("vendor(%s) in CR %s not supported for virtio-vdpa", iface.Vendor, policy.GetName())
 			}
+		} else {
+			glog.Infof("%s", err)
```

Comment:
So the logic before was if any interface selected was set to true (i.e. validateNicModel returned true) then the  print `fmt.Errorf("no supported NIC is selected by the nicSelector in CR %s", cr.GetName())` won't be activated. However with this new logic, we will be verbosely logging every interface that doesn't match. Do you have a sample log of this that you can share? And it would be apparent.

So we need to somehow store all the debug information going through all the interfaces such that when  `validateNicModel` never returned true then we will give the user all the debug information because nothing matched. This is to improve the verbosity of `no supported NIC is selected by the nicSelector in CR %s` print and not to add verbosity overall.

@bn222  Is this your understanding as well?

Reply:
yes, One approach (and please do give your feedback on this) is to call an additional function that only logs things. It would log which nics it tried and why it didn't work out.

Reply:
I think I am on the same page as you.

Reply:
@vrindle Removing this print does not solve the issue. There needs to be additional logic to consolidate the information and present it only in the case when no interfaces match.

Reply:
Right I am implementing that in a second commit. This one commit is only to ensure that the general cleanup is done.

Reply:
Can we rename the PR so that it's clear that we are doing a cleanup first and do all the refactoring in this first PR. The aim to reduce the size of the PR is that it can then go in. Next, we can have an additional PR after that that only adds the error message we wanted.

Reply:
That sounds good. I renamed the PR. This one only has the cleanup commit. Then there is this PR: https://github.com/k8snetworkplumbingwg/sriov-network-operator/pull/434

Please ensure this PR gets merged before merging the one above since this is a stacked PR.